File: elfa/a3_fast_api/app/services/llm.py
from typing import Any, Dict, AsyncGenerator
import httpx
from tenacity import retry, stop_after_attempt, wait_exponential

from app.core.config import settings
from app.core.logging import logger


class LLMClient:
    """Cliente genérico para integração com LLMs"""
    
    def __init__(self, provider: str | None = None):
        self.provider = provider or settings.llm_provider
        self.timeout = settings.llm_timeout
        self.max_retries = settings.llm_max_retries
        
        logger.info(f"LLM Client inicializado: {self.provider}")
    
    @retry(
        stop=stop_after_attempt(3),
        wait=wait_exponential(multiplier=1, min=2, max=10)
    )
    async def invoke(
        self,
        prompt: str,
        options: Dict[str, Any] | None = None
    ) -> str:
        """
        Invoca o LLM com um prompt.
        Implementação específica do provider deve ser adicionada.
        """
        logger.info(
            "LLM invocation",
            extra={
                "provider": self.provider,
                "prompt_length": len(prompt)
            }
        )
        
        # Implementação específica do provider aqui
        # Exemplo: Bedrock, OpenAI, etc.
        raise NotImplementedError(f"Provider {self.provider} não implementado")
    
    async def stream(
        self,
        prompt: str,
        options: Dict[str, Any] | None = None
    ) -> AsyncGenerator[str, None]:
        """
        Invoca o LLM em modo streaming.
        Implementação específica do provider deve ser adicionada.
        """
        logger.info(
            "LLM streaming",
            extra={"provider": self.provider}
        )
        
        # Implementação de streaming aqui
        raise NotImplementedError("Streaming não implementado")
        
        # Exemplo de uso:
        # async for chunk in self.stream(prompt):
        #     yield chunk

File: elfa/elfa_backend/src/services/llm.js
import { logger } from '../core/logger.js';
import { config } from '../core/config.js';

export class LLMClient {
  constructor(provider = config.llm.provider) {
    this.provider = provider;
    this.timeout = config.llm.timeout * 1000;
  }

  async invoke(prompt, options = {}) {
    logger.info('LLM invocation', { provider: this.provider, prompt });
    
    // Implementação específica do provider deve ser adicionada aqui
    throw new Error(`Provider ${this.provider} não implementado`);
  }

  async stream(prompt, options = {}) {
    logger.info('LLM streaming', { provider: this.provider });
    
    // Implementação de streaming deve ser adicionada aqui
    throw new Error('Streaming não implementado');
  }
}
